### DISCUSSION TITLE
Why is torrent read lock at CLient level?

### DISCUSSION THREAD

**Discussion by vijayanandnandam** [anacrolix/torrent#547] [Q&A] [answered]
Any specific reason why lock is always at client level and not a torrent level for torrent specific functions? I am facing issues when downloading torrents with a queue size of >= 100. The client seems to hang after sometime. It always hangs when trying to execute, `torrent.BytesCompleted` function.

**Comment by anacrolix** [comment]
Mainly as a simplification. There are some features that operate across the whole client, such as storage prioritization and request handling, common trackers, global connection limits etc. If you provide specific examples, like the BytesCompleted method I can check in to it.

**Comment by anacrolix** [comment]
I didn't change the locking, but try `torrent@master`, I made a few tweaks that might improve things. Please send or share a CPU performance profile if you can.

**Comment by vijayanandnandam** [comment]
@anacrolix let me try it with the master. I am not sure how to share the CPU profile. Pls guide me. I can share the code in private gist if you want to take a look

**Reply by anacrolix** [reply]
With https://github.com/anacrolix/envpprof imported, run with `GOPPROF=http` and send the file generated after running `go pprof http://envpprof:handler/debug/pprof/profile` or `GOPPROF=cpu`, and its corresponding profile when the binary completes.

Feel free to share a gist with me, but if it's not easy to reproduce, the profiles might be better.

**Reply by vijayanandnandam** [reply]
Have sent the details to to your emails.

**Comment by vijayanandnandam** [comment]
There is huge contention for mutex lock. Let me know if you want a code walk through. If reading takes too much time. All it is doing is

* Get magnet from redis
* Add to client
* save torrent file once it got the info
* download all files
* Periodically check for download status of all torrents and stop downloaded ones.
* It drop torrents if it cannot fetch magnet link within 300s. Drop torrents that are completed

**Comment by anacrolix** [comment, marked as answer]
This issue has moved into a private email dialogue. The fix https://github.com/anacrolix/torrent/commit/16c4b0888510fe2467745d041f4e70625f4d3514 should directly address the described issue, and some other fixes resulted while working through it.

**Comment by vijayanandnandam** [comment]
We can close this discussion. The main problem is solved and fixed

**Reply by anacrolix** [reply]
I think you can mark https://github.com/anacrolix/torrent/discussions/547#discussioncomment-1204108 as an answer to "close" this discussion.
